tianshou/utils/space_info.py

```diff
@@ -0,0 +1,81 @@
+from collections.abc import Sequence
+from dataclasses import dataclass, field
+
+import numpy as np
+from gymnasium import spaces
+
+
+@dataclass(kw_only=True)
+class ActionSpaceInfo:
+    action_shape: int | Sequence[int]
+    action_dim: int = field(init=False)
```

Comment:
Could you pls add dostrings? Especially for this field, I still don't fully understand why it's needed.

You could also instead just add a method `get_action_dim` or a property that reads it it. The post-init thing is not really safe - a user could overwrite action_shape after instantiation and the instance would be "broken" then (action_dim out of sync with action_shape). It's a bit contrived, but still, no need to allow for this possibility

Reply:
You can inherit from `ToStringMixin` and use it to include properties are included in the `__str__` and `__repr__`

Reply:
>  still don't fully understand why it's needed.
You could also instead just add a method get_action_dim or a property that reads it it.

After looking more into how it is being used, I think we can remove `action_dim` and `obs_dim` and let the caller use the `action_shape`/`obs_shape` appropriately 

(Some algorithms like [ICM take a product of all dimensions](https://github.com/dantp-ai/tianshou/blob/7f81ac73370bdfe398ada927297b0eff84c72de4/test/modelbased/test_ppo_icm.py#L133), so the current definition in `SpaceInfo` is not quite correct, although I suspect that the number of dimensions there is 1 so it works out).

Reply:
Sounds good, feel free to remove it

Reply:
However, if we remove it, we still need to check in the caller's code if `action/obs_dim` is `int | Sequence[int]` to ensure correct type checking, otherwise indexing the first value will not always work.

My suggestion is that we go ahead and remove `action/obs_dim`, but make `action_shape/obs_shape` of type `Sequence[int]` only. Then the caller can define `action/obs_dim` however they need (take the first value, multiply the dimension, ...).

Reply:
However, if we remove it, we still need to check in the caller's code if `action/obs_dim` is `int | Sequence[int]` to ensure correct type checking, otherwise indexing the first value will not always work.

We could remove `action/obs_dim`, but make `action_shape/obs_shape` of type `Sequence[int]` only. Then the caller can define `action/obs_dim` however they need (take the first value, multiply the dimension, ...).

Other networks accept either `int | Sequence[int]` (`Net`) or `Sequence[int]` (`Actor`) so it won't break APIs

Reply:
What is the semantics here? I guess it currently only supports discrete or a flat Box, in which case the dim is `action_shape[0]`. For a general Box, I suppose the correct semantics is the product of the dimensions.  We can keep it, but better as a method (`get_action_dim`) that returns either `.n` or the product of dims. What do you think?

Reply:
Fixed this, and added doc strings. I think we are very close to use this util to fix the issue on space info typing that is present in all the other test files.

Reply:
Structure looks good to me now, thanks! I added some minor comments, feel free to "roll out" this solution to all test-files and examples after addressing them